spec/requests/users_controller_spec.rb

```diff
@@ -5899,6 +5900,194 @@ def create_totp
+
+    it "fails if no user is logged in" do
+      post "/u/rename_passkey/NONE.json"
+
+      expect(response.status).to eq(403)
+    end
+
+    it "fails if no name parameter is provided" do
+      sign_in(user1)
+      post "/u/rename_passkey/ID.json"
+
+      expect(response.status).to eq(400)
+      expect(response.parsed_body["errors"][0]).to eq(
+        "param is missing or the value is empty: name",
+      )
+    end
+
+    it "fails if key is not present" do
+      sign_in(user1)
+      post "/u/rename_passkey/ID.json", params: { name: "new name" }
+
+      expect(response.status).to eq(400)
+      expect(response.parsed_body["errors"][0]).to include("Discourse::InvalidParameters")
+    end
+
+    context "with an existing passkey" do
+      fab!(:passkey) { Fabricate(:passkey_with_random_credential, user: user1) }
+
+      it "renames the key" do
+        sign_in(user1)
+        post "/u/rename_passkey/#{passkey.id}.json", params: { name: "new name" }
+        response_parsed = response.parsed_body
+
+        expect(response.status).to eq(200)
+        expect(passkey.reload.name).to eq("new name")
+      end
+    end
+  end
+
+  describe "#delete_passkey" do
+    before { SiteSetting.experimental_passkeys = true }
+
+    it "fails if user is not logged in" do
+      stub_secure_session_confirmed
+      delete "/u/delete_passkey/ID.json"
+      expect(response.status).to eq(403)
+    end
+
+    it "fails if session is not confirmed" do
+      sign_in(user1)
+      post "/u/register_passkey.json"
+      expect(response.status).to eq(403)
+    end
+
+    context "with a confirmed session" do
+      before { stub_secure_session_confirmed }
+
+      it "fails if key is not present" do
+        sign_in(user1)
+        delete "/u/delete_passkey/12.json"
+        expect(response.status).to eq(400)
+        expect(response.parsed_body["errors"][0]).to include("Discourse::InvalidParameters")
+      end
+
+      context "if the user has a passkey" do
+        fab!(:passkey) { Fabricate(:passkey_with_random_credential, user: user1) }
+
+        it "works" do
+          sign_in(user1)
+          delete "/u/delete_passkey/#{passkey.id}.json"
+          expect(response.status).to eq(200)
+          expect(user1.passkey_credential_ids).to eq([])
+        end
+
+        it "does not let another user delete a passkey associated with another user" do
+          sign_in(admin)
+          delete "/u/delete_passkey/#{passkey.id}.json"
+          expect(response.status).to eq(400)
+          expect(response.parsed_body["errors"][0]).to include("Discourse::InvalidParameters")
+        end
+      end
+    end
+  end
+
+  describe "#register_passkey" do
+    before { SiteSetting.experimental_passkeys = true }
+
+    it "fails if user is not logged in" do
+      stub_secure_session_confirmed
+      post "/u/register_passkey.json"
+
+      expect(response.status).to eq(403)
+    end
+
+    it "fails if session is not confirmed" do
```

Comment:
I'm not quite sure what session is not confirmed means. Will it be possible to clarify this?

Reply:
Yes, so adding a passkey is a sensitive action. It changes the user's relationship with their account, very similar to adding/deleting 2FA methods. So for these "sensitive actions", before letting the user make changes, we ask for an extra confirmation. Currently this is asking for password confirmation again. 

Other services, like github or google, let you make sensitive action changes by confirming either a password, a 2FA method or a passkey. 

<img width="533" alt="image" src="https://github.com/discourse/discourse/assets/368961/bbcf1e49-b48b-4ebb-9a6a-1818612beabf">

I think we should do the same, in a subsequent PR (would expand scope of this PR a bit too much). 

Not sure how else to call this other than "confirm session" or "trusted session". Open to ideas, of course.

Reply:
Got it :+1: Thank you for clarifying.